File: .skills/openclaw-skills/skills/0xnyk/xint/lib/action_result.ts
import type { ActionResultType } from "./actions";
import type { XintError } from "./errors";

export type ActionExecutionKind = Extract<ActionResultType, "success" | "error" | "info">;

export type Pagination = {
  total: number;
  returned: number;
  has_more: boolean;
};

export type ActionExecutionResult<T = unknown> = {
  type: ActionExecutionKind;
  message: string;
  data?: T;
  fallbackUsed: boolean;
  error?: XintError;
  cached?: boolean;
  cost?: number;
  pagination?: Pagination;
};

export function actionSuccess<T>(message: string, data?: T, fallbackUsed = false): ActionExecutionResult<T> {
  return { type: "success", message, data, fallbackUsed };
}

export function actionInfo<T>(message: string, data?: T, fallbackUsed = false): ActionExecutionResult<T> {
  return { type: "info", message, data, fallbackUsed };
}

export function actionError(message: string, error?: XintError): ActionExecutionResult<never> {
  return { type: "error", message, fallbackUsed: false, error };
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/action_result.rs
#[allow(dead_code)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ActionResultType {
    Success,
    Error,
    Info,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ActionExecutionResult<T> {
    pub result_type: ActionResultType,
    pub message: String,
    pub data: Option<T>,
    pub fallback_used: bool,
}

pub fn action_success<T>(message: impl Into<String>, data: Option<T>) -> ActionExecutionResult<T> {
    ActionExecutionResult {
        result_type: ActionResultType::Success,
        message: message.into(),
        data,
        fallback_used: false,
    }
}

#[allow(dead_code)]
pub fn action_info<T>(message: impl Into<String>, data: Option<T>) -> ActionExecutionResult<T> {
    ActionExecutionResult {
        result_type: ActionResultType::Info,
        message: message.into(),
        data,
        fallback_used: false,
    }
}

pub fn action_error<T>(message: impl Into<String>) -> ActionExecutionResult<T> {
    ActionExecutionResult {
        result_type: ActionResultType::Error,
        message: message.into(),
        data: None,
        fallback_used: false,
    }
}

#[cfg(test)]
mod tests {
    use super::{action_error, action_info, action_success, ActionResultType};

    #[test]
    fn constructors_set_result_type() {
        let success = action_success("ok", Some(1usize));
        assert_eq!(success.result_type, ActionResultType::Success);

        let info = action_info::<usize>("note", None);
        assert_eq!(info.result_type, ActionResultType::Info);

        let err = action_error::<usize>("fail");
        assert_eq!(err.result_type, ActionResultType::Error);
    }
}
